components/kyma-environment-broker/internal/ias/client.go

```diff
@@ -0,0 +1,231 @@
+package ias
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	kebError "github.com/kyma-incubator/compass/components/kyma-environment-broker/internal/error"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	PathServiceProviders  = "/service/sps"
+	PathCompanyGlobal     = "/service/company/global"
+	PathAccess            = "/service/sps/%s/rba"
+	PathIdentityProviders = "/service/idp"
+	PathDelete            = "/service/sps/delete"
+)
+
+type (
+	ClientConfig struct {
+		URL    string
+		ID     string
+		Secret string
+	}
+
+	Client struct {
+		config         ClientConfig
+		httpClient     *http.Client
+		closeBodyError error
+	}
+
+	Request struct {
+		Method  string
+		Path    string
+		Body    io.Reader
+		Headers map[string]string
+	}
+)
+
+func NewClient(cli *http.Client, cfg ClientConfig) *Client {
+	return &Client{
+		config:     cfg,
+		httpClient: cli,
+	}
+}
+
+func (c Client) CloseBodyError() error {
+	return c.closeBodyError
+}
+
+func (c *Client) SetType(spID string, payload Type) error {
+	return c.call(c.serviceProviderPath(spID), payload)
+}
+
+func (c *Client) SetAssertionAttribute(spID string, payload PostAssertionAttributes) error {
+	return c.call(c.serviceProviderPath(spID), payload)
+}
+
+func (c *Client) SetSubjectNameIdentifier(spID string, payload SubjectNameIdentifier) error {
+	return c.call(c.serviceProviderPath(spID), payload)
+}
+
+func (c *Client) SetAuthenticationAndAccess(spID string, payload AuthenticationAndAccess) error {
+	pathAccess := fmt.Sprintf(PathAccess, spID)
+
+	return c.call(pathAccess, payload)
+}
+
+func (c *Client) GetCompany() (*Company, error) {
+	company := &Company{}
+	request := &Request{Method: http.MethodGet, Path: PathCompanyGlobal}
+
+	response, err := c.do(request)
+	defer c.closeResponseBody(response)
```

Comment:
imo this needs to be replaced with:
```go
	defer func() {
		err = multierror.Append(err, errors.Wrap(c.closeResponseBody(response), "while trying to close body reader")).ErrorOrNil()
	}()
```

and `closeResponseBody` should  look like that:
```go
func (c *Client) closeResponseBody(response *http.Response) error {
	if response.Body == nil {
		return nil
	}
	return response.Body.Close()
}
```

as a result client does not have to play with calling another method like `CloseBodyError()`.,
```go
c, err := GetCompany()
if err != nil {
}
if err := c.CloseBodyError(); err != nil {
}
```
additionally, you also didnt use that method at all. What's more you didnt have any cleanup for CloseBodyError, so if such error occurred  once then it was always there which is quite misleading if you are executing e.g. `GetCompany` few times.